### PR TITLE
Add revisions for -Zlower-impl-trait-in-trait-to-assoc-ty fixed tests

### PR DESCRIPTION
**Pull request by spastorino**
Needs to go on top of #109198

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
Moving forward, can you not commit a separate "Add revisions to fixed tests" commit? If any of the first 3 commits fixes a UI test, then that UI test should be associated with each commit that fixes it. If the commit doesn't fix any UI tests, then it can probably be squashed together.

Basically I want to see--

Commit A
 * Changes some files
 * Fixes some UI tests

Commit B
  * Changes some files
  * Fixes some UI tests

It's hard to tell what is fixed by what, especially because there are not many comments in this PR really.

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 1]
```diff
@@ -198,9 +198,27 @@ fn check_static_inhabited(tcx: TyCtxt<'_>, def_id: LocalDefId) {
...
+            let trait_fn = tcx.impl_trait_in_trait_parent_fn(id.owner_id.to_def_id());
+            // Only check the validity of this opaque type if the function has a default body
+            if let hir::Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Fn(_, hir::TraitFn::Provided(_)),
+                ..
+            }) = tcx.hir().get_by_def_id(trait_fn.expect_local())
+            {
+                origin
+            } else {
+                return;
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/check.rs:212]: Why is this needed now?

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 2]
```diff
@@ -198,9 +198,27 @@ fn check_static_inhabited(tcx: TyCtxt<'_>, def_id: LocalDefId) {
 /// projections that would result in "inheriting lifetimes".
 fn check_opaque(tcx: TyCtxt<'_>, id: hir::ItemId) {
     let item = tcx.hir().item(id);
-    let hir::ItemKind::OpaqueTy(hir::OpaqueTy { origin, .. }) = item.kind else {
-        tcx.sess.delay_span_bug(item.span, "expected opaque item");
-        return;
+    let origin = match item.kind {
+        hir::ItemKind::OpaqueTy(hir::OpaqueTy { origin, in_trait: true, .. }) => {
+            let trait_fn = tcx.impl_trait_in_trait_parent_fn(id.owner_id.to_def_id());
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/check.rs:203]: You should be able to get this from the `origin`.

File: compiler/rustc_infer/src/infer/opaque_types.rs [hunk 3]
```diff
@@ -396,10 +396,29 @@ impl<'tcx> InferCtxt<'tcx> {
     /// defining scope.
     #[instrument(skip(self), level = "trace", ret)]
     fn opaque_type_origin_unchecked(&self, def_id: LocalDefId) -> OpaqueTyOrigin {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/opaque_types.rs:398]: It makes me uncomfortable that we're calling `opaque_type_origin_unchecked` on an associated type. That's probably wrong...

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 4]
```diff
@@ -2231,6 +2231,10 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         // ambiguous impls. The latter *ought* to be a
         // coherence violation, so we don't report it here.
 
+        if self.tcx.opt_rpitit_info(obligation.cause.body_id.to_def_id()).is_some() {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2234]: It's not generally sound to skip reporting an error here, without at least delaying a bug or something else. Can you explain why you're doing this?

**Comment by spastorino** [comment]
> Moving forward, can you not commit a separate "Add revisions to fixed tests" commit? If any of the first 3 commits fixes a UI test, then that UI test should be associated with each commit that fixes it. If the commit doesn't fix any UI tests, then it can probably be squashed together with the other commits that fully fix a test or not.

We need all the changes to make the working tests pass. Each commit was making this tests ICE in a new part of the code :).

File: compiler/rustc_hir_analysis/src/check/check.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_hir_analysis/src/check/check.rs:212]: Because we want to avoid checking the validity of the opaque if the trait has no default body.

File: compiler/rustc_infer/src/infer/opaque_types.rs (hunk 3 above)
Comment (spastorino) [reply compiler/rustc_infer/src/infer/opaque_types.rs:398]: Going to check why is being called.

File: compiler/rustc_hir_analysis/src/check/check.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/check.rs:212]: Of course. But we had opaque types before in trait methods with no default body. What *changed* to make it so that this code path is being called now?

Comment (spastorino) [reply compiler/rustc_hir_analysis/src/check/check.rs:212]: Ok, I see your point. Maybe it's more obvious if we do at this level as it's done in `ImplTraitPlaceholder` https://github.com/spastorino/rust/blob/e352207f21c8524f68f4dedb3ee7c55cf392b734/compiler/rustc_hir_analysis/src/check/check.rs#L587-L595

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/check.rs:212]: That link is broken, but yeah, I think I agree.

File: compiler/rustc_infer/src/infer/opaque_types.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/opaque_types.rs:398]: I think this is fixed alternatively by #109198, and should make it so that we don't need to modify `opaque_type_origin_unchecked` at all.

File: compiler/rustc_hir_analysis/src/check/check.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_hir_analysis/src/check/check.rs:212]: https://github.com/spastorino/rust/blob/master/compiler/rustc_hir_analysis/src/check/check.rs#L588-L598

File: compiler/rustc_infer/src/infer/opaque_types.rs (hunk 3 above)
Comment (spastorino) [reply compiler/rustc_infer/src/infer/opaque_types.rs:397]: Can you tell me how #109198 avoid calling this function?.

Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/opaque_types.rs:397]: Given some `-> impl Trait`, let's call that opaque `def-id#1` (`DefKind::Opaque`). We later synthesize a `def-id#2`, which is an associated type (`DefKind::AssocTy`).

In `typeck_with_fallback`, we normalize `ty::Alias(ty::Projection, def-id#2)` in the original trait signature to `ty::Alias(ty::Opaque, def-id#2)` *instead* of `ty::Alias(ty::Opaque, def-id#1)`:

https://github.com/rust-lang/rust/blob/c5c43407603599f4ff2b217cc09be2cc5f39967d/compiler/rustc_ty_utils/src/ty.rs#L271-L272

https://github.com/rust-lang/rust/blob/c5c43407603599f4ff2b217cc09be2cc5f39967d/compiler/rustc_ty_utils/src/ty.rs#L287-L295

That's wrong, because def-id#2 is an associated type.

That passes a signature with `ty::Alias(ty::Opaque, def-id#2)` to `check_fn`, which passes it to `replace_opaque_types_with_inference_vars`. That tries to call `opaque_type_origin` on `def-id#2` (reminder: an associated type) which ICEs.

That is what I fixed in #109198.

**Comment by spastorino** [comment]
@compiler-errors only the last commit is relevant. Maybe you can even fetch that one up and place it in https://github.com/rust-lang/rust/pull/109198 ?

**Comment by compiler-errors** [comment]
my pr already got rolled up into a rollup lol

r=me when it's landed

@bors rollup

**Comment by spastorino** [comment]
This thing is going to fail until #109198 is merged and we rebase on top of it.

**Comment by spastorino** [comment]
@bors r-

Something is odd here.

**Comment by spastorino** [comment]
Now yes ...

@bors r=compiler-errors
